### ISSUE TITLE
Squiz.Commenting.FunctionCommentThrowTag.Missing triggering incorrectly

### ISSUE THREAD

**Issue by timscaffidi** [squizlabs/PHP_CodeSniffer#633]
I have a method that internally throws an exception within a try/catch and catches it, and does not throw any external exceptions, like so:
```php
function wontThrow() {
    // do some safe things
    ...
    try {
        // do some dangerous things that may throw their own exceptions
        ...
        if ($some_additional_error_condition) {
            throw new Exception("An error condition has been met!");
        }
    } catch (Exception $e) {
        // safely handle the error state
        ...
    }
}
```

This is causing Squiz.Commenting.FunctionCommentThrowTag.Missing to trigger, however I don't think it should because the function itself is not throwing an exception, as it is internally caught.
Or am I misunderstanding something and is this how it is supposed to work?

**Comment by gsherwood** [comment]
That sniff isn't designed to account for code like you have written. I'm not entirely sure it is worth trying to detect it as I can't know if the catch is actually catching that specific exception type (or a parent type) or if it is catching some exceptions in the TRY and letting others be caught higher up. So I don't think I can make it accurate.

Is this something you do often or is this a one-off that the sniff is confused about?

**Comment by timscaffidi** [comment]
I probably do this sort of thing quite often. Is it bad form to throw and catch your own exception within a single function (Not being facetious, I really do not know)? I like to do this when there are multiple error conditions that I would like to handle in a similar fashion, for example in the case of a REST controller which would need to return a 500 code if anything breaks unexpectedly. Rather than having multiple return points I can catch them all and handle the error response once if each one just throws an exception.

**Comment by aik099** [comment]
I don't recommend using exceptions for control flow related tasks.

**Comment by timscaffidi** [comment]
How is what I described just 'control flow related' and not error handling? I detected an error state and throw an exception. It just so happens that the try and catch are in the same function. What's the difference between throwing an exception directly within a try block versus throwing one within a function that is called from within a try block? Is the former merely 'control flow' and the latter 'error handling'? Are exceptions not handling the flow of execution?

**Comment by aik099** [comment]
I recommend doing try/catch at higher abstraction layer. Then you'll have:

* class that does the work and doesn't throw an exception 
* class that is rest friendly and throws what it needs 
* class that is Web friendly and throws what it needs

**Comment by gsherwood** [comment]
> I probably do this sort of thing quite often.

That doesn't leave many options then. You might not be able to use that sniff (or this particular error message). Or, you could change it to a warning and then check manually if PHPCS flags it.

The only other thing I can think of is documenting that you are throwing an exception, but that you are catching it internally, but that does feel like a waste of time given that logic is encapsulated in the function itself.

**Comment by timscaffidi** [comment]
> you could change it to a warning and then check manually if PHPCS flags it

I will just do this. Thank you
